server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,626 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.apache.lucene.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.DocValuesConsumer;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.index.compositeindex.datacube.DateDimension;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.compositeindex.datacube.Metric;
+import org.opensearch.index.compositeindex.datacube.MetricStat;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeField;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeFieldConfiguration;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.MetricStatFieldPair;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregator;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregatorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.numerictype.StarTreeNumericType;
+import org.opensearch.index.compositeindex.datacube.startree.builder.SingleTreeBuilder;
+import org.opensearch.index.compositeindex.datacube.startree.builder.StarTreeDocValuesIteratorAdapter;
+import org.opensearch.index.compositeindex.datacube.startree.data.StarTreeDocument;
+import org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeBuilderUtils;
+import org.opensearch.index.fielddata.IndexNumericFieldData;
+import org.opensearch.index.mapper.Mapper;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.NumberFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for star-tree builder
+ */
+public abstract class BaseStarTreeBuilder implements SingleTreeBuilder {
```

Comment:
Can you add experimental tag to all public classes for now, we can later remove it once we come out of experimental mode ?

Reply:
Would not be able add ```@ExperimentalAPI``` annotation because NumericType from Lucene does not have this annotation. The way this annotation works is, it mandates the particular annotation to other referenced classes as well.
Will be adding ``` * @opensearch.experimental```` to the javadocs. That being said, wherever I can, will be adding ```@ExperimentalApi```